Zend/zend_compile.c

```diff
@@ -2222,6 +2233,91 @@ static zend_op *zend_delayed_compile_end(uint32_t offset) /* {{{ */
 }
 /* }}} */
 
+#define ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_WAS_ACTIVE (1 << 0)
+#define ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_STARTED (1 << 1)
+
+static zend_short_circuiting_scope *zend_current_short_circuiting_scope()
+{
+	return zend_stack_top(&CG(short_circuiting_scopes));
+}
+
+static zend_bool zend_ast_kind_is_short_circuited(uint32_t ast_kind)
+{
+	switch (ast_kind) {
+		case ZEND_AST_DIM:
+		case ZEND_AST_PROP:
+		case ZEND_AST_NULLSAFE_PROP:
+		case ZEND_AST_STATIC_PROP:
+		case ZEND_AST_METHOD_CALL:
+		case ZEND_AST_NULLSAFE_METHOD_CALL:
+		case ZEND_AST_STATIC_CALL:
+		case ZEND_AST_ASSIGN:
+		case ZEND_AST_ASSIGN_REF:
+		case ZEND_AST_ASSIGN_OP:
+		case ZEND_AST_ASSIGN_COALESCE:
+		case ZEND_AST_POST_INC:
+		case ZEND_AST_POST_DEC:
+		case ZEND_AST_PRE_INC:
+		case ZEND_AST_PRE_DEC:
+			return 1;
+		default:
+			return 0;
+	}
+}
+
+static uint32_t zend_begin_short_circuiting_scope(uint32_t ast_kind, zend_bool force)
+{
+	zend_bool short_circuiting_was_active = CG(short_circuiting);
+	zend_bool ast_kind_is_short_circuited = zend_ast_kind_is_short_circuited(ast_kind);
+
+	uint32_t flags = 0;
+	if (short_circuiting_was_active) {
+		flags |= ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_WAS_ACTIVE;
+	}
+
+	if (force || (!short_circuiting_was_active && ast_kind_is_short_circuited)) {
+		flags |= ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_STARTED;
+
+		CG(short_circuiting) = 1;
+
+		zend_short_circuiting_scope scope;
+		zend_stack_init(&scope.jmp_null_opnums, sizeof(uint32_t));
+		scope.flags = 0;
+		zend_stack_push(&CG(short_circuiting_scopes), &scope);
+	}
+
+	if (!force && !ast_kind_is_short_circuited) {
+		CG(short_circuiting) = 0;
+	}
+
+	return flags;
+}
+
+static void zend_end_short_circuiting_scope(uint32_t flags, znode *result)
+{
+	zend_bool short_circuiting_was_active = (flags & ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_WAS_ACTIVE) != 0;
+	zend_bool short_circuiting_started = (flags & ZEND_BEGIN_SHORT_CIRCUITING_SCOPE_STARTED) != 0;
+
+	if (short_circuiting_started) {
+		zend_short_circuiting_scope *scope = zend_current_short_circuiting_scope();
+		zend_stack *jmp_null_opnums = &scope->jmp_null_opnums;
+
+		while (!zend_stack_is_empty(jmp_null_opnums)) {
+			uint32_t jmp_null_opnum = *(uint32_t *)zend_stack_top(jmp_null_opnums);
+			zend_op *jmp_null_opline = &CG(active_op_array)->opcodes[jmp_null_opnum];
+			jmp_null_opline->op2.opline_num = get_next_op_number();
+			SET_NODE(jmp_null_opline->result, result);
+			jmp_null_opline->extended_value = scope->flags;
+			zend_stack_del_top(jmp_null_opnums);
+		}
+
+		zend_stack_del_top(&CG(short_circuiting_scopes));
+		zend_stack_destroy(jmp_null_opnums);
```

Comment:
Potential UAF?

Reply:
Haha I can't argue with that :P Which part?